examples_tests/50.RadixSort/main.cpp

```diff
@@ -14,6 +14,48 @@ struct SortElement { uint32_t key, data; };
 // Todo: There's a lot of room for optimization here, especially creating all the gpu resources again
 // and again when they can be created once and reused
 
+template <typename T>
+static T* DebugGPUBufferDownload(smart_refctd_ptr<IGPUBuffer> buffer_to_download, size_t buffer_size, IVideoDriver* driver)
+{
+	constexpr uint64_t timeout_ns = 15000000000u;
+	const uint32_t alignment = uint32_t(sizeof(T));
+	auto downloadStagingArea = driver->getDefaultDownStreamingBuffer();
+	auto downBuffer = downloadStagingArea->getBuffer();
+
+	bool success = false;
+
+	uint32_t array_size_32 = uint32_t(buffer_size);
+	uint32_t address = std::remove_pointer<decltype(downloadStagingArea)>::type::invalid_address;
+	auto unallocatedSize = downloadStagingArea->multi_alloc(1u, &address, &array_size_32, &alignment);
+	if (unallocatedSize)
+	{
+		os::Printer::log("Could not download the buffer from the GPU!", ELL_ERROR);
+		exit(420);
+	}
+
+	driver->copyBuffer(buffer_to_download.get(), downBuffer, 0, address, array_size_32);
+
+	auto downloadFence = driver->placeFence(true);
+	auto result = downloadFence->waitCPU(timeout_ns, true);
+
+	T* dataFromBuffer = nullptr;
+	if (result != video::E_DRIVER_FENCE_RETVAL::EDFR_TIMEOUT_EXPIRED && result != video::E_DRIVER_FENCE_RETVAL::EDFR_FAIL)
+	{
+		if (downloadStagingArea->needsManualFlushOrInvalidate())
+			driver->invalidateMappedMemoryRanges({ {downloadStagingArea->getBuffer()->getBoundMemory(),address,array_size_32} });
+
+		dataFromBuffer = reinterpret_cast<T*>(reinterpret_cast<uint8_t*>(downloadStagingArea->getBufferPointer()) + address);
+	}
+	else
+	{
+		os::Printer::log("Could not download the buffer from the GPU, fence not signalled!", ELL_ERROR);
+	}
+
+	downloadStagingArea->multi_free(1u, &address, &array_size_32, nullptr);
+
+	return dataFromBuffer;
+}
```

Comment:
doesn't this already exist in the engine in some form?

Reply:
no clue, just a quick thing I stole from example 48

this function won't be there in the final code, but this download code might